unix target check

JonathanWoollett-Light opened the pull request:
## Reason for This PR

Currently when attempting to build with a non-unix target you get numerous errors relating to `std::os::unix` not being available.

This is obtuse and annoying.

## Description of Changes

To simply inform the user why these errors are occurring a minimal `build.rs` can be added, which produces a simple error quickly when attempting to build on a non-unix system:
```
PS C:\Users\jcawl\Projects\micro-http> cargo build
   Compiling micro_http v0.1.0 (C:\Users\jcawl\Projects\micro-http)
error: This crate only supports unix targets.
 --> build.rs:3:5
  |
3 |     std::compile_error!("This crate only supports unix targets.");
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: could not compile `micro_http` due to previous error
PS C:\Users\jcawl\Projects\micro-http>
```

As opposed to the previous repetitive messages:
```
PS C:\Users\jcawl\Projects\micro-http> cargo build
   Compiling micro_http v0.1.0 (C:\Users\jcawl\Projects\micro-http)
error[E0433]: failed to resolve: could not find `unix` in `os`
 --> src\connection.rs:7:14
  |
7 | use std::os::unix::io::FromRawFd;
  |              ^^^^ could not find `unix` in `os`

error[E0433]: failed to resolve: could not find `unix` in `os`
 --> src\server.rs:6:14
  |
6 | use std::os::unix::io::{AsRawFd, FromRawFd, RawFd};       
  |              ^^^^ could not find `unix` in `os`

error[E0433]: failed to resolve: could not find `unix` in `os`
 --> src\server.rs:7:14
  |
7 | use std::os::unix::net::{UnixListener, UnixStream};
  |              ^^^^ could not find `unix` in `os`

error[E0432]: unresolved import `vmm_sys_util::sock_ctrl_msg`
  --> src\connection.rs:16:19
   |
16 | use vmm_sys_util::sock_ctrl_msg::ScmSocket;
   |                   ^^^^^^^^^^^^^ could not find `sock_ctrl_msg` in `vmm_sys_util`
error[E0432]: unresolved import `vmm_sys_util::sock_ctrl_msg`
  --> src\server.rs:15:19
   |
15 | use vmm_sys_util::sock_ctrl_msg::ScmSocket;
   |                   ^^^^^^^^^^^^^ could not find `sock_ctrl_msg` in `vmm_sys_util`
error[E0432]: unresolved import `vmm_sys_util::epoll`
  --> src\server.rs:17:5
   |
17 | use vmm_sys_util::epoll;
   |     ^^^^^^^^^^^^^^^^^^^ no `epoll` in the root

error[E0433]: failed to resolve: use of undeclared type `UnixListener`
   --> src\server.rs:276:22
    |
276 |         let socket = UnixListener::bind(path_to_socket).map_err(ServerError::IOError)?;
    |                      ^^^^^^^^^^^^ use of undeclared type `UnixListener`        

error[E0433]: failed to resolve: use of undeclared type `UnixListener`
   --> src\server.rs:297:31
    |
297 |         let socket = unsafe { UnixListener::from_raw_fd(socket_fd) };
    |                               ^^^^^^^^^^^^ use of undeclared type `UnixListener`

error[E0422]: cannot find struct, variant or union type `iovec` in crate `libc`      
   --> src\connection.rs:174:33
    |
174 |         let mut iovecs = [libc::iovec {
    |                                 ^^^^^ not found in `libc`

error[E0412]: cannot find type `UnixListener` in this scope
   --> src\server.rs:255:13
    |
255 |     socket: UnixListener,
    |             ^^^^^^^^^^^^ not found in this scope

error[E0412]: cannot find type `RawFd` in this scope
   --> src\server.rs:263:26
    |
253 | pub struct HttpServer {
    |                      - help: you might be missing a type parameter: `<RawFd>`  
...
263 |     connections: HashMap<RawFd, ClientConnection<UnixStream>>,
    |                          ^^^^^ not found in this scope

error[E0412]: cannot find type `UnixStream` in this scope
   --> src\server.rs:263:50
    |
253 | pub struct HttpServer {
    |                      - help: you might be missing a type parameter: `<UnixStream>`
...
263 |     connections: HashMap<RawFd, ClientConnection<UnixStream>>,
    |                                                  ^^^^^^^^^^ not found in this scope

error[E0412]: cannot find type `RawFd` in this scope
   --> src\server.rs:296:35
    |
296 |     pub fn new_from_fd(socket_fd: RawFd) -> Result<Self> {
    |                                   ^^^^^ not found in this scope

error[E0412]: cannot find type `RawFd` in this scope
   --> src\server.rs:555:36
    |
555 |                     response.id as RawFd,
    |                                    ^^^^^ not found in this scope

error[E0412]: cannot find type `RawFd` in this scope
   --> src\server.rs:603:51
    |
603 |     fn epoll_mod(epoll: &epoll::Epoll, stream_fd: RawFd, evset: epoll::EventSet) -> Result<()> {
    |                                                   ^^^^^ not found in this scope
error[E0412]: cannot find type `RawFd` in this scope
   --> src\server.rs:614:51
    |
614 |     fn epoll_add(epoll: &epoll::Epoll, stream_fd: RawFd) -> Result<()> {
    |                                                   ^^^^^ not found in this scope
error[E0412]: cannot find type `RawFd` in this scope
   --> src\server.rs:628:51
    |
628 |     fn epoll_del(epoll: &epoll::Epoll, stream_fd: RawFd) -> Result<()> {
    |                                                   ^^^^^ not found in this scope
Some errors have detailed explanations: E0412, E0422, E0432, E0433.
For more information about an error, try `rustc --explain E0412`.
error: could not compile `micro_http` due to 17 previo
```

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided (issue no. or explanation).
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.

mattschlebusch commented:
Summary of thoughts as follows:

* The objective for the CR is fairly reasonable, are there other packages/crates in Firecracker that may benefit from clarifying build errors related to the target OS?
    * On a similar vein, I think the README would benefit from a `supported and tested platforms` section to formalize this rule as well.
*  To try be a bit more precise and accurate with regards to the build error message,
    *  Unix is a very specific operating system. Were you intending to refer to all Unix "style" operating systems, `*nix*` more generally speaking?
    * Is it fair AND accurate to say this crate is intended for the Linux OS only?
* Avoid code in the overview of the PR, *especially* if it is a duplicate of the diff!
* It doesn't need to be a comprehensive copy-paste, but it would be nice to see an example of the type of exception that would have been generated before this change to gauge how much clarity the new error message is providing.

JonathanWoollett-Light commented:
> - The objective for the CR is fairly reasonable, are there other packages/crates in Firecracker that may benefit from clarifying build errors related to the target OS?
>    - On a similar vein, I think the README would benefit from a supported and tested platforms section to formalize this rule as well.

I differently agree on this as a wider approach.

> - To try be a bit more precise and accurate with regards to the build error message,
>   - Unix is a very specific operating system. Were you intending to refer to all Unix "style" operating systems, *nix* more generally speaking?
>   - Is it fair AND accurate to say this crate is intended for the Linux OS only?

I was trying to give a similar message to what you might see in [the Rust documentation](https://doc.rust-lang.org/stable/std/os/unix/index.html) where it illustrates:

![image](https://user-images.githubusercontent.com/16856679/169531680-a3691c9a-127b-4b80-8965-24603a26ac7e.png)

With this referring to the [`target_family = "unix"`](https://doc.rust-lang.org/reference/conditional-compilation.html#target_family), so not specifically Linux in this context but rather like the `Unix "style" operating systems` you mention. 

(edited my original pr to contains the previous exception and remove the new code)

mattschlebusch commented:
>> The objective for the CR is fairly reasonable, are there other packages/crates in Firecracker that may benefit from clarifying build errors related to the target OS?
>>  *  On a similar vein, I think the README would benefit from a supported and tested platforms section to formalize this rule as well.

> I differently agree on this as a wider approach.

To try be more concise, 
* Is there a place where the gap in build error messaging in other packages is being tracked?
* Is there any issue with updating the README.md here?

> I was trying to give a similar message to what you might see in [the Rust documentation](https://doc.rust-lang.org/stable/std/os/unix/index.html) where it illustrates

I understand the specificity of the documentation, this is due to the namespaces they are derived from. However if we are going to be putting in a package level build error we should be more accurate.

JonathanWoollett-Light commented:
> Is there a place where the gap in build error messaging in other packages is being tracked?

I'm not sure what you mean by this. Could you elaborate?

> Is there any issue with updating the README.md here?

I don't see any, I think we should also update `README.md`.

> we should be more accurate.

I'm not sure how to be more accurate here, what would you rather see the error message say?

mattschlebusch commented:
>> Is there a place where the gap in build error messaging in other packages is being tracked?

> I'm not sure what you mean by this. Could you elaborate?

What are the other crates/packages where this same issue occurs, and are there existing github issues to track fixing them there as well.

>> Is there any issue with updating the README.md here?

> I don't see any, I think we should also update README.md.

I don't see any changes to the README.md file. Did you forget a commit perhaps?

>> we should be more accurate.

> I'm not sure how to be more accurate here, what would you rather see the error message say?

Something like the following comes to mind:
* `This crate only supports *nix targets.`
* `This crate only supports Linux/Unix targets.`

JonathanWoollett-Light commented:
> What are the other crates/packages where this same issue occurs, and are there existing github issues to track fixing them there as well.

I don't know, perhaps there should be a wider issue/ticket created for implementing similar changes or considering a solution for all crates in the ecosystem, but I think this falls considerably outside the scope of this PR.

> I don't see any changes to the README.md file. Did you forget a commit perhaps?

I understand now you where asking me by implication to make these changes, I have updated the readme.

> - This crate only supports *nix targets.
> - This crate only supports Linux/Unix targets

How about `This crate only supports Unix-like targets`? (https://en.wikipedia.org/wiki/Unix-like)

mattschlebusch reviewed (dismissed):
I think there is still a misnomer here. The validation is not related to Unix. At the very least the error is aggregated for future devs to better understand.

Commit:
Clarifying unix-like support

Added `build.rs` to  produce a compile error when attempting to build on a
non-Unix-like system, and clarified support in `readme.md`.

Signed-off-by: Jonathan Woollett-Light <jcawl@amazon.co.uk>
